File: DeleteANodeInBst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class DeleteANodeInBst {
    // Searching a node in tree + Inorder Successor in Binary Search Tree
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root==null){
            return root; // return null
        }

        if(root.val <key){
            root.right = deleteNode(root.right, key);
            return root;
        } else if(root.val>key){
            root.left = deleteNode(root.left, key);
            return root;
        } else {
            // root.val == key
            // leaf node
            if(root.left == null && root.right ==null){
                return null;
            }
            if(root.left ==null){
                return root.right; // left child is null
            } else if(root.right == null){
                return root.left; // right child is null
            } else {
                // left child is not null and right child is not null
                TreeNode nextMinNode = root.right;
                while(nextMinNode.left !=null){
                    nextMinNode = nextMinNode.left;
                }

                nextMinNode.left = root.left;
                return root.right;
            }
        }
    }
}

File: November2021/C++/Delete_a_Node_in_BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root) return NULL;
        if(root->val<key)
            root->right= deleteNode(root->right,key);
        else if(root->val>key)
            root->left= deleteNode(root->left,key);
        else
        {
            if(!root->right)
                return root->left;
            else if(!root->left)
                return root->right;
            
            TreeNode* minnode=findmind(root->right);
            root->val=minnode->val;
            root->right=deleteNode(root->right,minnode->val);
        }
            return root;
    
    }
    
    TreeNode* findmind(TreeNode* node)
    {
        
        while((node->left))
            node=node->left;
        return node;
    }
};